mtp_api/apps/credit/views_admin.py

```diff
@@ -0,0 +1,123 @@
+from django import forms
+from django.db import models
+from django.db.models.functions import TruncMonth
+from django.utils.translation import gettext_lazy as _
+
+from core.forms import BasePeriodAdminReportForm, BasePrisonAdminReportForm
+from core.views import BaseAdminReportView
+from credit.models import Credit, CREDIT_RESOLUTION
+from prison.models import Prison
+
+
+class CreditReportForm(BasePeriodAdminReportForm):
+    private_estate = forms.ChoiceField(label=_('Private estate'), choices=(
+        ('include', _('Include')),
+        ('exclude', _('Exclude')),
+    ), initial='include')
+
+
+class CreditReportAdminView(BaseAdminReportView):
+    title = _('Credit report')
+    template_name = 'admin/credit/credit/report.html'
+    form_class = CreditReportForm
+
+    def get_context_data(self, **kwargs):
+        context_data = super().get_context_data(**kwargs)
+        form = context_data['form']
+
+        exclude_private_estate = form.cleaned_data['private_estate'] == 'exclude'
+
+        current_period = form.current_period
+        format_date = form.period_formatter
+
+        rows = form.group_months_into_periods(self.get_monthly_rows(exclude_private_estate))
+        rows = list(filter(lambda r: r['date'] < current_period, rows))
+        for row in rows:
+            row['date_label'] = format_date(row['date'])
+        context_data['rows'] = rows
+
+        context_data['opts'] = Credit._meta
+        context_data['form'] = form
+        context_data['exclude_private_estate'] = exclude_private_estate
+        return context_data
+
+    def get_monthly_rows(self, exclude_private_estate=False):
+        queryset = Credit.objects \
+            .filter(resolution=CREDIT_RESOLUTION.CREDITED) \
+            .exclude(prison__in=self.excluded_nomis_ids) \
+            .order_by()
+        if exclude_private_estate:
+            queryset = queryset.filter(prison__private_estate=False)
+        return queryset \
+            .annotate(date=TruncMonth('received_at')) \
+            .values('date') \
+            .annotate(count=models.Count('pk'), amount=models.Sum('amount')) \
+            .values('date', 'count', 'amount') \
+            .order_by('date')
+
+
+class PrisonCreditReportForm(BasePrisonAdminReportForm):
+    ordering = forms.ChoiceField(choices=(
+        ('nomis_id', _('Prison')),
+        ('count', _('Digital credits')),
+        ('amount', _('Digital credit amount')),
+    ), initial='nomis_id')
+
+
+class PrisonCreditReportAdminView(BaseAdminReportView):
+    title = _('Credits per-prison')
+    template_name = 'admin/credit/credit/prison-report.html'
+    form_class = PrisonCreditReportForm
+
+    def get_context_data(self, **kwargs):
+        context_data = super().get_context_data(**kwargs)
+        form = context_data['form']
+
+        prisons = Prison.objects \
+            .exclude(nomis_id__in=self.excluded_nomis_ids) \
+            .order_by() \
```

Comment:
it removes any default ordering. here it doesn't matter since the results are not used in any order. but in queries resulting in `GROUP BY` clauses, it makes a huge difference

Reply:
https://docs.djangoproject.com/en/2.0/topics/db/aggregation/#interaction-with-default-ordering-or-order-by